torch/csrc/Exceptions.cpp

```diff
@@ -137,5 +137,81 @@ ValueError::ValueError(const char *format, ...) {
   va_end(fmt_args);
 }
 
+// ATen warning handler for Python
+using warning_buffer_t =
+  std::vector<std::pair<c10::SourceLocation, std::string>>;
+
+static warning_buffer_t warning_buffer;
+static std::mutex warning_buffer_mutex;
+
+static void py_warning_handler(
+    const c10::SourceLocation& source_location,
+    const std::string& msg) {
+  std::unique_lock<std::mutex> lock(warning_buffer_mutex);
+  warning_buffer.push_back({source_location, msg});
+};
+
+EnforceWarningBuffer::EnforceWarningBuffer() noexcept(true): prev_handler(c10::Warning::get_warning_handler()) {
+  c10::Warning::set_warning_handler(&py_warning_handler);
+}
+
+/// See NOTE [ Conversion Cpp Python Warning ] for noexcept justification
+/// NOLINTNEXTLINE(bugprone-exception-escape)
+EnforceWarningBuffer::~EnforceWarningBuffer() noexcept(false) {
+  c10::Warning::set_warning_handler(prev_handler);
+
+  std::unique_lock<std::mutex> lock(warning_buffer_mutex);
+
+  if(warning_buffer.size() > 0) {
+    AutoGIL gil;
```

Comment:
Can `py_warning_handler` ever be called when we have the GIL? In that case, there's potential for deadlock here, right? One thread has the GIL and is waiting on warning buffer mutex, the other thread has warning buffer mutex and is waiting on the GIL.